File: linux-test-client/src/crypto/keys.ts
import { ed25519 } from '@noble/curves/ed25519';
import { randomBytes } from 'crypto';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

/**
 * Manages Ed25519 keypairs for E2E encryption
 */
export class KeyManager {
  private privateKey: Uint8Array | null = null;
  private publicKey: Uint8Array | null = null;
  private keyPath: string;

  constructor() {
    // Store keys in ~/.config/utter-client/keypair.key
    const configDir = path.join(os.homedir(), '.config', 'utter-client');
    if (!fs.existsSync(configDir)) {
      fs.mkdirSync(configDir, { recursive: true });
    }
    this.keyPath = path.join(configDir, 'keypair.key');
  }

  /**
   * Get or generate Ed25519 keypair
   */
  public getOrGenerateKeyPair(): void {
    if (fs.existsSync(this.keyPath)) {
      console.log('[Crypto] Loading existing keypair');
      this.loadKeyPair();
    } else {
      console.log('[Crypto] Generating new Ed25519 keypair');
      this.generateAndSaveKeyPair();
    }
  }

  /**
   * Generate new Ed25519 keypair and save to file
   */
  private generateAndSaveKeyPair(): void {
    // Generate Ed25519 keypair
    const privateKeyBytes = randomBytes(32);
    this.privateKey = new Uint8Array(privateKeyBytes);
    this.publicKey = ed25519.getPublicKey(this.privateKey);

    // Save private key to file
    fs.writeFileSync(this.keyPath, Buffer.from(this.privateKey));

    // Set restrictive permissions (Unix only)
    if (process.platform !== 'win32') {
      fs.chmodSync(this.keyPath, 0o600); // rw------- (owner only)
    }

    console.log('[Crypto] Keypair generated and saved to', this.keyPath);
  }

  /**
   * Load keypair from file
   */
  private loadKeyPair(): void {
    const keyBytes = fs.readFileSync(this.keyPath);

    if (keyBytes.length !== 32) {
      throw new Error(`Invalid key length: ${keyBytes.length} bytes (expected 32)`);
    }

    this.privateKey = new Uint8Array(keyBytes);
    this.publicKey = ed25519.getPublicKey(this.privateKey);
  }

  /**
   * Get the public key in base64 format
   */
  public getPublicKeyBase64(): string {
    if (!this.publicKey) {
      throw new Error('No keypair loaded. Call getOrGenerateKeyPair() first.');
    }
    return Buffer.from(this.publicKey).toString('base64');
  }

  /**
   * Get the private key bytes
   */
  public getPrivateKeyBytes(): Uint8Array {
    if (!this.privateKey) {
      throw new Error('No keypair loaded');
    }
    return this.privateKey;
  }

  /**
   * Get the public key bytes
   */
  public getPublicKeyBytes(): Uint8Array {
    if (!this.publicKey) {
      throw new Error('No keypair loaded');
    }
    return this.publicKey;
  }

  /**
   * Clear all stored keys (delete key file)
   */
  public clearKeys(): void {
    if (fs.existsSync(this.keyPath)) {
      fs.unlinkSync(this.keyPath);
      console.log('[Crypto] Keys cleared');
    }
    this.privateKey = null;
    this.publicKey = null;
  }
}

File: utterd/src/crypto/keys.rs
use ed25519_dalek::{SigningKey, VerifyingKey};
use rand::rngs::OsRng;
use std::fs;
use std::path::PathBuf;

/// Manages Ed25519 keypairs for E2E encryption
///
/// Keys are stored in ~/.config/utterd/keypair.key
pub struct KeyManager {
    config_dir: PathBuf,
    signing_key: Option<SigningKey>,
    verifying_key: Option<VerifyingKey>,
}

impl KeyManager {
    /// Create a new KeyManager with default config directory
    pub fn new() -> Result<Self, Box<dyn std::error::Error>> {
        let config_dir = dirs::config_dir()
            .ok_or("Could not find config directory")?
            .join("utterd");

        // Create config directory if it doesn't exist
        fs::create_dir_all(&config_dir)?;

        Ok(Self {
            config_dir,
            signing_key: None,
            verifying_key: None,
        })
    }

    /// Get or generate Ed25519 keypair
    pub fn get_or_generate_keypair(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let key_path = self.config_dir.join("keypair.key");

        if key_path.exists() {
            println!("[Crypto] Loading existing keypair from {:?}", key_path);
            self.load_keypair(&key_path)?;
        } else {
            println!("[Crypto] Generating new Ed25519 keypair");
            self.generate_and_save_keypair(&key_path)?;
        }

        Ok(())
    }

    /// Generate new Ed25519 keypair and save to file
    fn generate_and_save_keypair(&mut self, path: &PathBuf) -> Result<(), Box<dyn std::error::Error>> {
        let mut csprng = OsRng;
        let signing_key = SigningKey::generate(&mut csprng);
        let verifying_key = signing_key.verifying_key();

        // Save private key to file
        fs::write(path, signing_key.to_bytes())?;

        // Set restrictive permissions (Unix only)
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let mut perms = fs::metadata(path)?.permissions();
            perms.set_mode(0o600); // rw------- (owner only)
            fs::set_permissions(path, perms)?;
        }

        println!("[Crypto] Keypair generated and saved to {:?}", path);

        self.signing_key = Some(signing_key);
        self.verifying_key = Some(verifying_key);

        Ok(())
    }

    /// Load keypair from file
    fn load_keypair(&mut self, path: &PathBuf) -> Result<(), Box<dyn std::error::Error>> {
        let key_bytes = fs::read(path)?;

        if key_bytes.len() != 32 {
            return Err(format!("Invalid key length: {} bytes (expected 32)", key_bytes.len()).into());
        }

        let key_array: [u8; 32] = key_bytes.try_into()
            .map_err(|_| "Failed to convert key bytes to array")?;

        let signing_key = SigningKey::from_bytes(&key_array);
        let verifying_key = signing_key.verifying_key();

        self.signing_key = Some(signing_key);
        self.verifying_key = Some(verifying_key);

        Ok(())
    }

    /// Get the public key in base64 format
    pub fn get_public_key_base64(&self) -> Result<String, Box<dyn std::error::Error>> {
        let verifying_key = self.verifying_key
            .as_ref()
            .ok_or("No keypair loaded. Call get_or_generate_keypair() first.")?;

        Ok(base64::encode(verifying_key.as_bytes()))
    }

    /// Get the private key bytes
    pub fn get_private_key_bytes(&self) -> Result<&[u8; 32], Box<dyn std::error::Error>> {
        let signing_key = self.signing_key
            .as_ref()
            .ok_or("No keypair loaded")?;

        Ok(signing_key.as_bytes())
    }

    /// Get the public key bytes
    pub fn get_public_key_bytes(&self) -> Result<&[u8; 32], Box<dyn std::error::Error>> {
        let verifying_key = self.verifying_key
            .as_ref()
            .ok_or("No keypair loaded")?;

        Ok(verifying_key.as_bytes())
    }

    /// Clear all stored keys (delete key file)
    pub fn clear_keys(&self) -> Result<(), Box<dyn std::error::Error>> {
        let key_path = self.config_dir.join("keypair.key");
        if key_path.exists() {
            fs::remove_file(key_path)?;
            println!("[Crypto] Keys cleared");
        }
        Ok(())
    }
}

impl Default for KeyManager {
    fn default() -> Self {
        Self::new().expect("Failed to create KeyManager")
    }
}
